File: typescript/tests/cross-sdk-parity.test.ts
// Cross-SDK parity test for the TypeScript SDK.
//
// Consumes the canonical fixture corpus at <repo>/fixtures/ produced by
// python/tests/test_cross_sdk_parity.py. Asserts the TS SDK round-trips
// events / tasks and produces pricing output that matches the Python-
// canonical expected outputs.
//
// This suite is intentionally RED on initial commit. Each failing
// sub-test pins an audit finding scheduled for Sprint 1+:
//   - P1  occurred_at timestamp format drift
//   - P2  PricingSource enum spelling drift
//   - P4  network event 4xx-below-threshold emission
//   - B11 streaming-body corruption (downstream)
//   - LLM cost map drift (TS missing ~708 keys vs Python catalog)
//   - URL scrubber absent in TS (Theme A, Sprint 1)
//
// As each finding lands the corresponding sub-test must flip green.

import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { resolve, dirname, basename, join } from "node:path";
import { fileURLToPath } from "node:url";

import {
  eventFromDict,
  eventToDict,
  taskFromDict,
  taskToDict,
  PricingEngine,
} from "../src/index.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const FIXTURES_ROOT = resolve(__dirname, "..", "..", "fixtures");

function readJson(path: string): Record<string, unknown> {
  return JSON.parse(readFileSync(path, "utf-8"));
}

function stripUnderscoredKeys(d: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(d)) {
    if (k.startsWith("_")) continue;
    if (v !== null && typeof v === "object" && !Array.isArray(v)) {
      out[k] = stripUnderscoredKeys(v as Record<string, unknown>);
    } else {
      out[k] = v;
    }
  }
  return out;
}

function expectedPathFor(rel: string, kind: string): string {
  // Map fixture-input rel paths to their expected-output counterparts.
  //   events/foo.json                 -> canonical_serialization/foo.json
  //   events/edge_cases/foo.json      -> canonical_serialization/edge_cases/foo.json
  //   tasks/foo.json                  -> canonical_serialization/tasks/foo.json
  //   pricing_inputs/X/foo.json       -> pricing/X/foo.json
  const base = basename(rel);
  if (rel.startsWith("pricing_inputs/")) {
    return join(FIXTURES_ROOT, "expected_outputs", kind, rel.slice("pricing_inputs/".length));
  }
  if (rel.includes("edge_cases/")) {
    return join(FIXTURES_ROOT, "expected_outputs", kind, "edge_cases", base);
  }
  if (rel.startsWith("tasks/")) {
    return join(FIXTURES_ROOT, "expected_outputs", kind, "tasks", base);
  }
  return join(FIXTURES_ROOT, "expected_outputs", kind, base);
}

const EVENT_FIXTURES = [
  "events/llm_call.v1.json",
  "events/external_cost.v1.json",
  "events/compute_cost_lambda.v1.json",
  "events/compute_cost_ec2_share.v1.json",
  "events/compute_cost_k8s_pod.v1.json",
  "events/network.v1.json",
  "events/network_4xx_below_threshold.v1.json",
  "events/gpu_cost.v1.json",
  "events/gpu_utilization_signal.v1.json",
  "events/retry_marker.v1.json",
  "events/edge_cases/tiny_decimal.v1.json",
];

const TASK_FIXTURES = ["tasks/task_minimal.v1.json", "tasks/task_with_network_gpu.v1.json"];

describe("cross-SDK event canonical serialization", () => {
  for (const rel of EVENT_FIXTURES) {
    it(rel, () => {
      const input = stripUnderscoredKeys(readJson(join(FIXTURES_ROOT, rel)));
      const expected = readJson(expectedPathFor(rel, "canonical_serialization"));
      const evt = eventFromDict(input);
      // Round-trip through JSON to normalize Date/number/Decimal types into
      // the same shape as the expected file.
      const actual = JSON.parse(JSON.stringify(eventToDict(evt)));
      expect(actual).toEqual(expected);
    });
  }
});

describe("cross-SDK task canonical serialization", () => {
  for (const rel of TASK_FIXTURES) {
    it(rel, () => {
      const input = stripUnderscoredKeys(readJson(join(FIXTURES_ROOT, rel)));
      const expected = readJson(expectedPathFor(rel, "canonical_serialization"));
      const task = taskFromDict(input);
      const actual = JSON.parse(JSON.stringify(taskToDict(task)));
      expect(actual).toEqual(expected);
    });
  }
});

describe("cross-SDK LLM pricing parity", () => {
  const engine = new PricingEngine();
  const LLM_FIXTURES = [
    "pricing_inputs/llm/gpt4o_500_in_200_out.json",
    "pricing_inputs/llm/claude_sonnet_streaming_2000_in_1500_out.json",
  ];
  for (const rel of LLM_FIXTURES) {
    it(rel, () => {
      const input = stripUnderscoredKeys(readJson(join(FIXTURES_ROOT, rel))) as Record<
        string,
        number | string
      >;
      const expected = readJson(expectedPathFor(rel, "pricing"));
      const result = engine.getCost(
        input.model as string,
        input.input_tokens as number,
        input.output_tokens as number,
        (input.cached_tokens as number) ?? 0,
        0,
      );
      // Decimal equality (not string equality) — TS uses number, so trailing
      // zeros / repr differ from Python's Decimal stringification. Drift in
      // the numeric value itself (cost-map differences) is the real signal.
      expect(Number(result.costUsd)).toBeCloseTo(Number(expected.cost_usd), 8);
      expect(result.pricingSource).toBe(expected.pricing_source);
    });
  }
});

describe.skip("cross-SDK URL scrubber parity", () => {
  // Sprint 1 / Theme A: TS SDK has no URL scrubber. Pin the gap.
  // expected_outputs/security/url_with_*.v1.json defines the canonical algorithm.
  it("TODO(sprint-1, theme-a): implement scrubUrl() in TS SDK security module", () => {});
});

describe("cross-SDK tiny-decimal accumulation invariant (B3)", () => {
  it("1.23E-8 summed 10000 times equals 0.0001230000 exactly", () => {
    const expected = readJson(
      join(FIXTURES_ROOT, "expected_outputs", "pricing", "decimal_accumulation_invariant.json"),
    );
    const per = Number(expected.per_event_cost_usd);
    const iters = Number(expected.iterations);
    const wantTotal = Number(expected.total_cost_usd);

    let total = 0;
    for (let i = 0; i < iters; i++) {
      total += per;
    }
    // RED today: native JS number arithmetic drifts. B3 fix in Sprint 2
    // (decimal.js everywhere) flips this green.
    expect(total).toBe(wantTotal);
  });
});

File: go/cross_sdk_parity_test.go
package dexcost

// Cross-SDK parity test for the Go SDK.
//
// Consumes the canonical fixture corpus at /home/user/dexcost-sdk/fixtures/
// (relative path: ../fixtures/) produced by python/tests/test_cross_sdk_parity.py.
// Asserts the Go SDK round-trips events / tasks and produces pricing output
// byte-equal (canonical_serialization) or decimal-equal (pricing) to the
// Python-canonical expected outputs.
//
// This test is intentionally RED on initial commit. Each failing sub-test
// pins an audit finding scheduled for Sprint 1+:
//   - B5  ec2 / k8s_pod compute discriminator
//   - B6  gpu_cost / gpu_utilization_signal schema enum gap
//   - P1  occurred_at timestamp format drift (Go RFC3339Nano vs Python +00:00)
//   - P2  PricingSource enum spelling drift
//   - URL scrubber absent in Go (Theme A, Sprint 1)
//
// As each finding is fixed the corresponding sub-test must flip green.

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"

	"github.com/DexwoxBusiness/dexcost-go/core"
	"github.com/DexwoxBusiness/dexcost-go/pricing"
	"github.com/DexwoxBusiness/dexcost-go/security"
	"github.com/shopspring/decimal"
)

var fixturesRoot = filepath.Join("..", "fixtures")

func readJSONFile(t *testing.T, path string) map[string]interface{} {
	t.Helper()
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	var out map[string]interface{}
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatalf("parse %s: %v", path, err)
	}
	return out
}

func stripUnderscoredKeys(d map[string]interface{}) map[string]interface{} {
	cleaned := make(map[string]interface{}, len(d))
	for k, v := range d {
		if strings.HasPrefix(k, "_") {
			continue
		}
		if nested, ok := v.(map[string]interface{}); ok {
			cleaned[k] = stripUnderscoredKeys(nested)
		} else {
			cleaned[k] = v
		}
	}
	return cleaned
}

func normalizeViaJSON(t *testing.T, v interface{}) map[string]interface{} {
	t.Helper()
	raw, err := json.Marshal(v)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var out map[string]interface{}
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	return out
}

func expectedPathFor(rel string, kind string) string {
	// Map fixture-input relative paths to their expected-output counterparts.
	//   events/foo.json                       -> canonical_serialization/foo.json
	//   events/edge_cases/foo.json            -> canonical_serialization/edge_cases/foo.json
	//   tasks/foo.json                        -> canonical_serialization/tasks/foo.json
	//   pricing_inputs/{compute|gpu|egress|llm}/foo.json -> pricing/{compute|gpu|egress|llm}/foo.json
	base := filepath.Base(rel)
	switch {
	case strings.HasPrefix(rel, "pricing_inputs/"):
		// strip "pricing_inputs/" prefix, keep subdir
		sub := strings.TrimPrefix(rel, "pricing_inputs/")
		return filepath.Join(fixturesRoot, "expected_outputs", kind, sub)
	case strings.Contains(rel, "edge_cases/"):
		return filepath.Join(fixturesRoot, "expected_outputs", kind, "edge_cases", base)
	case strings.HasPrefix(rel, "tasks/"):
		return filepath.Join(fixturesRoot, "expected_outputs", kind, "tasks", base)
	default:
		return filepath.Join(fixturesRoot, "expected_outputs", kind, base)
	}
}

var eventFixtures = []string{
	"events/llm_call.v1.json",
	"events/external_cost.v1.json",
	"events/compute_cost_lambda.v1.json",
	"events/compute_cost_ec2_share.v1.json",
	"events/compute_cost_k8s_pod.v1.json",
	"events/network.v1.json",
	"events/network_4xx_below_threshold.v1.json",
	"events/gpu_cost.v1.json",
	"events/gpu_utilization_signal.v1.json",
	"events/retry_marker.v1.json",
	"events/edge_cases/tiny_decimal.v1.json",
}

var taskFixtures = []string{
	"tasks/task_minimal.v1.json",
	"tasks/task_with_network_gpu.v1.json",
}

func TestCrossSDKEventCanonicalSerialization(t *testing.T) {
	for _, rel := range eventFixtures {
		rel := rel
		t.Run(rel, func(t *testing.T) {
			input := stripUnderscoredKeys(readJSONFile(t, filepath.Join(fixturesRoot, rel)))
			expected := readJSONFile(t, expectedPathFor(rel, "canonical_serialization"))

			evt, err := core.EventFromDict(input)
			if err != nil {
				t.Fatalf("EventFromDict(%s): %v", rel, err)
			}
			actual := normalizeViaJSON(t, evt.ToDict())

			if !reflect.DeepEqual(actual, expected) {
				actualJSON, _ := json.MarshalIndent(actual, "", "  ")
				expectedJSON, _ := json.MarshalIndent(expected, "", "  ")
				t.Errorf("event canonical-serialization drift for %s\n--- expected ---\n%s\n--- actual ---\n%s",
					rel, expectedJSON, actualJSON)
			}
		})
	}
}

func TestCrossSDKTaskCanonicalSerialization(t *testing.T) {
	for _, rel := range taskFixtures {
		rel := rel
		t.Run(rel, func(t *testing.T) {
			input := stripUnderscoredKeys(readJSONFile(t, filepath.Join(fixturesRoot, rel)))
			expected := readJSONFile(t, expectedPathFor(rel, "canonical_serialization"))

			task, err := core.TaskFromDict(input)
			if err != nil {
				t.Fatalf("TaskFromDict(%s): %v", rel, err)
			}
			actual := normalizeViaJSON(t, task.ToDict())

			if !reflect.DeepEqual(actual, expected) {
				actualJSON, _ := json.MarshalIndent(actual, "", "  ")
				expectedJSON, _ := json.MarshalIndent(expected, "", "  ")
				t.Errorf("task canonical-serialization drift for %s\n--- expected ---\n%s\n--- actual ---\n%s",
					rel, expectedJSON, actualJSON)
			}
		})
	}
}

func decimalEqual(t *testing.T, expected, actual interface{}, label string) {
	t.Helper()
	expStr, ok := expected.(string)
	if !ok {
		t.Fatalf("%s: expected value not a string: %T %v", label, expected, expected)
	}
	expDec, err := decimal.NewFromString(expStr)
	if err != nil {
		t.Fatalf("%s: parse expected %q: %v", label, expStr, err)
	}
	var actDec decimal.Decimal
	switch v := actual.(type) {
	case decimal.Decimal:
		actDec = v
	case string:
		actDec, err = decimal.NewFromString(v)
		if err != nil {
			t.Fatalf("%s: parse actual %q: %v", label, v, err)
		}
	default:
		t.Fatalf("%s: actual type unsupported: %T %v", label, actual, actual)
	}
	if !expDec.Equal(actDec) {
		t.Errorf("%s drift: expected=%s actual=%s", label, expDec.String(), actDec.String())
	}
}

func TestCrossSDKLLMPricingParity(t *testing.T) {
	engine, err := pricing.NewEngine()
	if err != nil {
		t.Fatalf("NewEngine: %v", err)
	}
	for _, rel := range []string{
		"pricing_inputs/llm/gpt4o_500_in_200_out.json",
		"pricing_inputs/llm/claude_sonnet_streaming_2000_in_1500_out.json",
	} {
		rel := rel
		t.Run(rel, func(t *testing.T) {
			input := stripUnderscoredKeys(readJSONFile(t, filepath.Join(fixturesRoot, rel)))
			expected := readJSONFile(t, expectedPathFor(rel, "pricing"))

			model, _ := input["model"].(string)
			inTok := int(input["input_tokens"].(float64))
			outTok := int(input["output_tokens"].(float64))
			cached := 0
			if c, ok := input["cached_tokens"].(float64); ok {
				cached = int(c)
			}

			result := engine.GetCost(model, inTok, outTok, cached, 0)

			decimalEqual(t, expected["cost_usd"], result.CostUSD, rel+":cost_usd")
			if got, want := result.PricingSource, expected["pricing_source"]; got != want {
				t.Errorf("%s pricing_source drift: expected=%v actual=%v", rel, want, got)
			}
		})
	}
}

func TestCrossSDKURLScrubberParity(t *testing.T) {
	// Sprint 1 / Theme A. Compares security.ScrubURL byte-for-byte against
	// expected_outputs/security/url_with_*.v1.json (Python-canonical).
	urlFixtures := []string{
		"events/edge_cases/url_with_basic_auth.v1.json",
		"events/edge_cases/url_with_api_key_query.v1.json",
		"events/edge_cases/url_with_signed_s3.v1.json",
	}
	for _, rel := range urlFixtures {
		rel := rel
		t.Run(rel, func(t *testing.T) {
			input := readJSONFile(t, filepath.Join(fixturesRoot, rel))
			testInput, ok := input["_test_input"].(map[string]interface{})
			if !ok {
				t.Fatalf("%s: missing _test_input", rel)
			}
			rawURL, _ := testInput["url"].(string)
			expected := readJSONFile(t, filepath.Join(fixturesRoot,
				"expected_outputs", "security", filepath.Base(rel)))

			actualScrubbed := security.ScrubURL(rawURL)
			expectedScrubbed, _ := expected["scrubbed_url"].(string)
			expectedRaw, _ := expected["raw_url"].(string)
			if expectedRaw != rawURL {
				t.Errorf("%s: raw_url mismatch fixture=%q expected=%q", rel, rawURL, expectedRaw)
			}
			if actualScrubbed != expectedScrubbed {
				t.Errorf("%s: scrub drift\n  raw:      %s\n  expected: %s\n  actual:   %s",
					rel, rawURL, expectedScrubbed, actualScrubbed)
			}
		})
	}
}

// TestCrossSDKTinyDecimalAccumulation pins the B3 invariant:
// summing 1.23E-8 ten thousand times must equal 0.0001230000 exactly.
func TestCrossSDKTinyDecimalAccumulation(t *testing.T) {
	expected := readJSONFile(t, filepath.Join(fixturesRoot, "expected_outputs", "pricing", "decimal_accumulation_invariant.json"))
	per := decimal.RequireFromString(expected["per_event_cost_usd"].(string))
	iters := int(expected["iterations"].(float64))
	wantTotal := decimal.RequireFromString(expected["total_cost_usd"].(string))

	var total decimal.Decimal
	for i := 0; i < iters; i++ {
		total = total.Add(per)
	}
	if !total.Equal(wantTotal) {
		t.Errorf("decimal accumulation drift: %d * %s = %s, expected %s",
			iters, per.String(), total.String(), wantTotal.String())
	}
}
